File: andriodGL/app/src/main/java/z3d/frame3d/Frame3dRes.java
package z3d.frame3d;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import org.json.JSONException;
import org.json.JSONObject;

import scene.CallBack;
import z3d.base.ByteArray;
import z3d.base.CallBackFun;
import z3d.base.ObjData;
import z3d.base.Scene_data;
import z3d.md5.Md5Analysis;
import z3d.md5.MeshImportSort;
import z3d.md5.MeshToObjUtils;
import z3d.res.BaseRes;
import z3d.scene.Scene3D;
import z3d.units.LoadBackFun;
import z3d.units.LoadManager;

public class Frame3dRes extends BaseRes {
    private static final String TAG ="Frame3dRes" ;
//public  static float  frameNum=1;
    private CallBack _completeFun;
    public void load(String url, CallBack bfun){
        _completeFun=bfun;
        LoadManager.getInstance().loadUrl(Scene_data.fileRoot + url,LoadManager.BYTE_TYPE, new LoadBackFun() {
            @Override
            public void bfun(HashMap dic) {
                if(dic!=null){
                    ByteArray temp=(ByteArray)dic.get("byte");
                    loadComplete(temp);
                }else{

                }
            }
        },null);
    }
    public static int frameSpeedNum;
    public static String sceneFileroot;
    public static String fileName;
    public boolean haveVideo;
    public void loadComplete(ByteArray $byte) {
        this._byte =  $byte;
        this.version = this._byte.readInt();
        String $str  = this._byte.readUTF();
        String[] itemstr=  $str.split("/");
        Frame3dRes.sceneFileroot = $str.replace(itemstr[itemstr.length - 1], "");
        Frame3dRes.fileName = itemstr[itemstr.length - 1];
        Frame3dRes.frameSpeedNum = this._byte.readInt();

        Log.d(TAG, "版本"+ this.version+ "frameSpeedNum"+ Frame3dRes.frameSpeedNum);

        this.readSceneInfo();


        this.read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext();
            }
        });


    }
    public void readNext() {
        this.read();//obj
        this.read();//material
        this.read();//particle;
        this.readFrame3dScene();

    }
    public List<FrameNodeVo> frameItem;
    public void readFrame3dScene() {
        try {
            this.frameItem =new ArrayList<>();
            int size = this._byte.readInt();
            JSONArray $scene = new JSONArray(this._byte.readUTFBytes(size));

        for (int i = 0; i < $scene.length(); i++) {
            FrameNodeVo $frameNodeVo = new FrameNodeVo();
            $scene.get(i);
            $frameNodeVo.writeObject((JSONObject)$scene.get(i));
            this.frameItem.add($frameNodeVo);
        }
        this._completeFun.StateChange(true);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }

    private void readSceneInfo(){
        int size = this._byte.readInt();
        try {
            JSONObject $obj = new JSONObject(this._byte.readUTFBytes(size));
            this.haveVideo = $obj.getBoolean("haveVideo");
//            Scene_data.light.setData($obj.SunNrm, $obj.SunLigth, $obj.AmbientLight);
//            LightBmpModel.getInstance().videoLightUvData = $obj.videoLightUvData;

        } catch (JSONException e) {
            e.printStackTrace();
        }



    }




}

File: new/src/Pan3d/frame3d/Frame3dRes.ts
module Pan3d {

    export class Frame3dRes extends BaseRes {
        private _completeFun: Function;
        public load($url: string, $completeFun: Function): void {
            this._completeFun = $completeFun;
            LoadManager.getInstance().load($url, LoadManager.BYTE_TYPE, ($byte: ArrayBuffer) => {
                this.loadComplete($byte);
            }, null);
        }
        public static frameSpeedNum: number;
        public static sceneFileroot: string;
        public static fileName: string;
        public haveVideo: boolean
        public loadComplete($byte: ArrayBuffer): void {

            this._byte = new Pan3dByteArray($byte);
            this._byte.position = 0;
            this.version = this._byte.readInt();
            if (this.version >= 31) {

            }
            var $str = this._byte.readUTF();
            var $itemstr: Array<string> = $str.split("/")
            Frame3dRes.sceneFileroot = $str.replace($itemstr[$itemstr.length - 1], "")
            Frame3dRes.fileName = $itemstr[$itemstr.length - 1]


            Frame3dRes.frameSpeedNum = this._byte.readInt();
            console.log("版本", this.version, "frameSpeedNum", Frame3dRes.frameSpeedNum)


            this.readSceneInfo();


            this.read(() => { this.readNext() });//img
        }
        private toVect4($num: number): Vector3D {
            var temp: number = Math.floor(65536 * $num);
            var a: number = Math.floor(temp / 256);
            var b: number = Math.floor(temp - a * 256);
            return new Vector3D(a / 256, b / 256, 0, 1);

        }
        private toNum(vect: Vector3D): number {
            var $a: number = vect.x * 256;
            var $b: number = vect.y * 256;
            var $bnum: number = ($a * 256 + $b) / 65536
            console.log("$bnum", $bnum)
            return ($a * 256 + $b) / 65536
        }
        //收获环境参数
        private readSceneInfo(): void {
            var size: number = this._byte.readInt();
            var $obj: any = JSON.parse(this._byte.readUTFBytes(size));

            this.haveVideo = $obj.haveVideo;
            // Scene_data.light.setData($obj.SunNrm, $obj.SunLigth, $obj.AmbientLight);
            // LightBmpModel.getInstance().videoLightUvData = $obj.videoLightUvData;

        }
        public static frameNum: number = 1
        public readNext(): void {
            this.read();//obj
            this.read();//material
            this.read();//particle;
            this.readFrame3dScene()

        }
        public frameItem: Array<FrameNodeVo>
        public readFrame3dScene(): void {
            this.frameItem = new Array;
            var size: number = this._byte.readInt();
            var $scene = JSON.parse(this._byte.readUTFBytes(size));
            for (var i: number = 0; i < $scene.length; i++) {
                var $frameNodeVo: FrameNodeVo = new FrameNodeVo()
                $frameNodeVo.writeObject($scene[i])
                this.frameItem.push($frameNodeVo)
            }
            this._completeFun();
        }
    }

    export class FrameSceneChar extends SceneChar {
        private shadowShadr: Shader3D
        shadow: boolean;
        constructor(value: Scene3D) {
            super(value);

        }

    }



}